pkg/stream/deduplicate.go

```diff
@@ -39,38 +40,44 @@ func Deduplicate(in <-chan []byte, window time.Duration, ticker func(time.Durati
 	return out
 }
 
-type dedupe struct{ *btree.BTree }
+type dedupe struct {
+	*btree.BTree
+	ulidRecordmap map[WrapULID][]byte // map[ulid]record && no race
+}
 
 func (d dedupe) insert(record []byte) {
-	d.BTree.ReplaceOrInsert(item(record))
+	if len(record) <= ulid.EncodedSize {
+		panic("invalid record")
+	}
+	var node ulid.ULID
+	node.UnmarshalText(record[:ulid.EncodedSize])
+	d.BTree.ReplaceOrInsert(WrapULID{node})
+	d.ulidRecordmap[WrapULID{node}] = record
```

Comment:
When do you delete from the ulidRecordMap? It seems like it grows without bounds!